encoding/ssh/filexfer/packets.go

```diff
@@ -0,0 +1,227 @@
+		return new(RenamePacket), nil
+	case PacketTypeReadlink:
+		return new(ReadlinkPacket), nil
+	case PacketTypeSymlink:
+		return new(SymlinkPacket), nil
+	case PacketTypeExtended:
+		return new(ExtendedPacket), nil
+	default:
+		return nil, fmt.Errorf("unexpected request packet type: %v", typ)
+	}
+}
+
+// RawPacket implements the general packet format from draft-ietf-secsh-filexfer-02
+//
+// Defined in https://tools.ietf.org/html/draft-ietf-secsh-filexfer-02#section-3
+type RawPacket struct {
+	Type      PacketType
+	RequestID uint32
+
+	Data Buffer
+}
+
+// Reset clears the pointers and reference-semantic variables of RawPacket,
+// making it suitable to be put into a sync.Pool.
+func (p *RawPacket) Reset() {
+	p.Data.Reset()
+}
+
+// MarshalPacket returns p as a two-part binary encoding of p.
+//
+// The internal p.RequestID is overridden by the reqid argument.
+func (p *RawPacket) MarshalPacket(reqid uint32) (header, payload []byte, err error) {
+	b := NewMarshalBuffer(p.Type, reqid, 0)
+
+	return b.Packet(p.Data.Bytes())
+}
+
+// MarshalBinary returns p as the binary encoding of p.
+func (p *RawPacket) MarshalBinary() ([]byte, error) {
+	return ComposePacket(p.MarshalPacket(p.RequestID))
+}
+
+// UnmarshalFrom decodes a RawPacket from the given Buffer into p.
+func (p *RawPacket) UnmarshalFrom(buf *Buffer) error {
+	typ, err := buf.ConsumeUint8()
+	if err != nil {
+		return err
+	}
+
+	p.Type = PacketType(typ)
+
+	if p.RequestID, err = buf.ConsumeUint32(); err != nil {
+		return err
+	}
+
+	p.Data = *buf
+	return nil
+}
+
+// UnmarshalBinary decodes a full raw packet out of the given data.
+// It is assumed that the uint32(length) has already been consumed to receive the data.
+//
+// NOTE: To avoid extra allocations, UnmarshalBinary aliases the given byte slice.
+func (p *RawPacket) UnmarshalBinary(data []byte) error {
+	return p.UnmarshalFrom(NewBuffer(data))
+}
+
+// defaultMaxPacketSize is defined in draft-ietf-secsh-filexfer-02 section 3.
+const defaultMaxPacketSize = 34000
+
+// readPacket reads a uint32 length-prefixed binary data packet from r.
+// If the given buffer is less than 4-bytes, it allocates a new buffer of size defaultMaxPacketSize.
+func readPacket(r io.Reader, b []byte) ([]byte, error) {
+	if len(b) < 4 {
+		b = make([]byte, defaultMaxPacketSize)
+	}
+
+	if _, err := io.ReadFull(r, b[:4]); err != nil {
+		return nil, err
+	}
+
+	length := unmarshalUint32(b)
+	if length < 1 {
+		return nil, ErrShortPacket
+	}
+	if length > uint32(len(b)) {
+		return nil, ErrLongPacket
+	}
+
+	n, err := io.ReadFull(r, b[:length])
+	return b[:n], err
+}
+
+// ReadFrom reads a full raw packet out of the given reader.
+func (p *RawPacket) ReadFrom(r io.Reader, b []byte) error {
+	b, err := readPacket(r, b)
+	if err != nil {
+		return err
+	}
+
+	return p.UnmarshalBinary(b)
+}
+
+// RequestPacket decodes a full request packet from the internal Data based on the Type.
+func (p *RawPacket) RequestPacket() (*RequestPacket, error) {
```

Comment:
Given that a RequestPacket is only three words, this method could return by value. That way, RequestPacket.Reset is no longer needed because it doesn't have to be pooled.

Reply:
Just because you’re passing something by value does not mean that it does not also still get dragged into the heap. A lot of issues that I have been running into now are from things that _I_ can clearly see should be on the stack escaping to the heap because the compiler cannot be sure.

Reply:
Presumably, that happens only when you assign to an interface value (?). Those are pointers in disguise, even when you put a non-pointer value in them.

Reply:
Also, oddly, just because something is passed by pointer does not mean that the compiler has actually put it on the heap. Many might be surprised that `b := make([]byte, 4)` can allocated on the stack, or that a `r := &RequestPacket{}` can also be on the stack.

Meanwhile, even something like `fmt.Printf("constant format string: %d", i)` can have its arguments escape to heap. 🙃 It’s been a very _very_ annoying go around with this stuff, and confusing choices made by the compiler. I have generally found that any virtualized function call will force its receiver to escape to the heap even if its receiver is a value rather than pointer type, and the only way to prevent such a transfer to heap is if the whole chain can be inlined and the virtualized function can be devirtualized. At that point, it doesn’t matter if it is by value or receiver, it does not automatically escape to heap.

Also, any buffers that are subject to a `r.Read(…)` also escape to heap, even if we can patently see that it doesn’t really have to. But because the contents are handed off to a virtualized subfunction that could potentially retain the buffer, it has to be moved to the heap. 😭 

My point here is: switching types just in order to prevent escaping to heap is way more complicated than can be resolved just by switching from a pointer to value type. That said, I have looked at the potentials of having the `MarshalPacket` types be value receivers, as this seems to increase the likelihood of them devirtualizing… but I’m also pretty sure that I cannot prevent this, as we very likely will have to call some other virtualized method on the type, causing it to escape to heap even if it is constructed by value.

Reply:
I don’t think I addressed this in the other comment, but no, not just because it’s assigned into an interface value. If a struct is large enough, keeping it in the stack even once becomes a liability, and Go will instead allocate that value on the heap. Even though it’s a value.

Tracking what goes on the stack and what goes on the heap is simply not as simple as “it’s secretly a pointer” or not. Plenty of values that are interfaces slices, or even pointer values can avoid allocation on the heap.

Reply:
I can't reproduce that with Go 1.16:

```go
package main

import "fmt"

func main() {
	var x foo
	f(x)
}

type foo struct {
	a [1 << 20]byte
}

func f(x foo) {
	switch x.a[0] {
	}
	fmt.Println(x.a[0])
}
```

According to -gcflags=-m, f cannot be inlined. In the generated assembly, I see a copy of 131072 quadwords onto the stack and an increment of the stack pointer by a megabyte. The Println call then copies one byte to the heap :)

When I change the size of foo.a to 1<<30, compilation fails:

```
./main.go:4:6: stack frame too large (>1GB): 1024 MB locals + 0 MB args
./main.go:14:6: stack frame too large (>1GB): 0 MB locals + 1024 MB args
```

Reply:
Same with return values:

```go
//go:noinline
func f() [1<<20]byte {}
```

copies a megabyte to the stack, but with 1<<30 it fails to compile.